op-supervisor/supervisor/backend/db/db.go

```diff
@@ -31,58 +28,92 @@ type LogStorage interface {
-	Contains(blockNum uint64, logIdx uint32, logHash common.Hash) (nextIndex entrydb.EntryIdx, err error)
+	// Contains returns no error iff the specified logHash is recorded in the specified blockNum and logIdx.
+	// If the log is out of reach, then ErrFuture is returned.
+	// If the log is determined to conflict with the canonical chain, then ErrConflict is returned.
+	// logIdx is the index of the log in the array of all logs in the block.
+	// This can be used to check the validity of cross-chain interop events.
+	// The block-seal of the blockNum block, that the log was included in, is returned.
+	// This seal may be fully zeroed, without error, if the block isn't fully known yet.
+	Contains(blockNum uint64, logIdx uint32, logHash common.Hash) (includedIn types.BlockSeal, err error)
+}
+
+type LocalDerivedFromStorage interface {
+	Last() (derivedFrom eth.BlockRef, derived eth.BlockRef, err error)
+	AddDerived(derivedFrom eth.BlockRef, derived eth.BlockRef) error
+	LastDerived(derivedFrom eth.BlockID) (derived eth.BlockID, err error)
+	DerivedFrom(derived eth.BlockID) (derivedFrom eth.BlockID, err error)
+}
+
+type CrossDerivedFromStorage interface {
+	LocalDerivedFromStorage
+	// This will start to differ with reorg support
 }
 
 var _ LogStorage = (*logs.DB)(nil)
 
-// ChainsDB is a database that stores logs and heads for multiple chains.
+// ChainsDB is a database that stores logs and derived-from data for multiple chains.
 // it implements the ChainsStorage interface.
 type ChainsDB struct {
-	logDBs      map[types.ChainID]LogStorage
-	safetyIndex safety.SafetyIndex
-	logger      log.Logger
+	// RW mutex:
+	// Read = chains can be read / mutated.
+	// Write = set of chains is changing.
+	mu sync.RWMutex
+
+	// unsafe info: the sequence of block seals and events
+	logDBs map[types.ChainID]LogStorage
+
+	// cross-unsafe: how far we have processed the unsafe data.
+	crossUnsafe map[types.ChainID]types.BlockSeal
+
+	// local-safe: index of what we optimistically know about L2 blocks being derived from L1
+	localDBs map[types.ChainID]LocalDerivedFromStorage
+
+	// cross-safe: index of L2 blocks we know to only have cross-L2 valid dependencies
+	crossDBs map[types.ChainID]CrossDerivedFromStorage
+
+	// finalized: the L1 finality progress. This can be translated into what may be considered as finalized in L2.
+	// It is initially zeroed, and the L2 finality query will return
+	// an error until it has this L1 finality to work with.
+	finalizedL1 eth.L1BlockRef
+
+	logger log.Logger
 }
 
-func NewChainsDB(logDBs map[types.ChainID]LogStorage, l log.Logger) *ChainsDB {
-	ret := &ChainsDB{
-		logDBs: logDBs,
-		logger: l,
+func NewChainsDB(l log.Logger) *ChainsDB {
+	return &ChainsDB{
+		logDBs:      make(map[types.ChainID]LogStorage),
+		logger:      l,
+		localDBs:    make(map[types.ChainID]LocalDerivedFromStorage),
+		crossDBs:    make(map[types.ChainID]CrossDerivedFromStorage),
+		crossUnsafe: make(map[types.ChainID]types.BlockSeal),
```

Comment:
microscopic nit: could we get all the `make(map`s grouped up
```suggestion
		logDBs:      make(map[types.ChainID]LogStorage),
		localDBs:    make(map[types.ChainID]LocalDerivedFromStorage),
		crossDBs:    make(map[types.ChainID]CrossDerivedFromStorage),
		crossUnsafe: make(map[types.ChainID]types.BlockSeal),
		logger:      l,		
```